### ISSUE TITLE
What makes a frame janky?

### ISSUE THREAD

**Issue by kenzieschmoll** [flutter/devtools#451]
Starting an issue to track this discussion: **what makes a frame janky**?

Right now, we consider a frame to be janky when the sum of UI and GPU durations exceed the target time limit per frame (currently hard coded to ~16ms). We do not give each thread a 16ms threshold because we are also concerned with latency.

**Question**: To determine whether a frame is janky, should we A) _add the durations of UI work and GPU work_ (current implementation), or B) _use the duration of the frame_?

Case 1:
When both the sum of durations and frame duration meet the target, it does not matter which way we determine jank.
![Screen Shot 2019-03-20 at 2 44 37 PM](https://user-images.githubusercontent.com/43759233/54721279-c5bb4f80-4b1e-11e9-90c6-e50cacb87829.png)

Case 2 (problem case):
The sum of durations exceeds the target, but the frame duration does not.
![Screen Shot 2019-03-20 at 2 44 46 PM](https://user-images.githubusercontent.com/43759233/54721286-cd7af400-4b1e-11e9-9a9a-28659e49b0d8.png)

Case 3 (problem case):
The frame duration exceeds the target, but the sum of durations does not.
![Screen Shot 2019-03-20 at 2 44 53 PM](https://user-images.githubusercontent.com/43759233/54721302-db307980-4b1e-11e9-99b0-9a65c0c0072e.png)

**Comment by kenzieschmoll** [comment]
@chinmaygarde @liyuqian @Hixie

**Comment by liyuqian** [comment]
Good question. I think a clear definition would help us and our developers a lot.

My personal understanding is that "jank" is the opposite of "smooth". Latency is another issue. It's kind like when you buy a gaming monitor, you'll check 2 separate specs:

1. Refresh rate, which is about the smoothness. Often monitors have 60fps but 144hz is now quite popular for first-person shooting gamers.
2. Input latency. It could have a 240hz refresh rate but a high latency (say 10ms), or a 60hz refresh rate with low latency (say 1ms).

I think the separation of smoothness and latency also applies to Flutter. It's best to measure, monitor, and optimize them separately.

For the "jank" or smoothness specifically, it's just about how continuous your frames are. If you have two frames that differ from one another significantly in an unexpected way, users will detect such "jank" or non-smoothness.

In my experience, there are 2 ways to have such non-smoothness:
1. There's too much a time gap between the first frame and the second frame. It's usually caused by the first frame taking up too much time which creates an discontinuity in time, and thus a discontinuity in pixels (assuming that each pixel is a smooth function in time). Be aware that the the time gap may not be equal to the total time that 1st frame takes when there's concurrency. For example, in Flutter where 2 threads exist, 1st frame could take 32ms to render while we only maintain a 16ms time gap between 2 frames.
2. The pixel itself is unexpectedly non-smooth in time. This is usually caused by a bug in the application or the framework which unexpectedly turns a smooth physical process into a non-smooth animation. In this case, no matter how fast a frame renders, the users will still detect the "jank". See https://github.com/flutter/flutter/issues/813#issuecomment-403946266 and https://github.com/flutter/flutter/issues/19222 for an example.

Now let's turn back to the latency. When there's a high latency, users may still see a very smooth animation, but they'll feel that the animation is lagged behind their input. Note that "an input event" is required for a nontrivial latency issue. If there's no user input and you see a lagged but smooth animation, you can simply shift your time function a little to synchronize everything. So if we're going to measure the latency, it's better to measure from the start of the input event. Measuring from the start of rendering is Ok (just like the monitors only measure latency from the input signal to the output pixels). But be aware that doesn't tell the full story of user-experienced latency.

**Comment by xuning0** [comment]
@kenzieschmoll I have some doubts about the case 2 diagram. In my understanding, after the cpu calculation is completed, data is submitted to the gpu for rendering. Why does the gpu start rendering before the cpu ends in the diagram? Thank you    for your answer.

**Comment by TheVinhLuongFossil** [comment]
Let's say I have two subsequent frames A and A+1 in chronological order. Is it possible for frame A is still in the middle of rasterizing while frame A+1 is still building (raster time of A overlaps with frame A+1's build time)?

**Comment by kenzieschmoll** [comment]
> Is it possible for frame A is still in the middle of rasterizing while frame A+1 is in building phase (raster time of A overlaps with frame A+1's build time)?

Yes this is possible, and most of the time expected. Build and Raster happen on two different threads and can happen in parallel.

**Comment by kenzieschmoll** [comment]
Closing since there is no action to be taken here.
